heat/core/tiling.py

```diff
@@ -0,0 +1,756 @@
+        are not generated above the diagonal in the case that gshape[0] << gshape[1].
+
+        WARNING: The generation of these tiles may unbalance the original tensor!
+
+        Note: This tiling scheme is intended for use with the QR function.
+
+        Parameters
+        ----------
+        arr : DNDarray
+            the array to be tiled
+        tiles_per_proc : int, optional
+            Default = 2
+            the number of divisions per process,
+
+        Returns
+        -------
+        None
+
+        Initializes
+        -----------
+        __col_per_proc_list : list
+            list is length of the number of processes, each element has the number of tile
+            columns on the process whos rank equals the index
+        __DNDarray = arr : DNDarray
+            the whole DNDarray
+        __lshape_map : torch.Tensor
+            unit -> [rank, row size, column size]
+            tensor filled with the shapes of the local tensors
+        __tile_map : torch.Tensor
+            units -> row, column, start index in each direction, process
+            tensor filled with the global indices of the generated tiles
+        __row_per_proc_list : list
+            list is length of the number of processes, each element has the number of tile
+            rows on the process whos rank equals the index
+        """
+        # lshape_map -> rank (int), lshape (tuple of the local lshape, self.lshape)
+        if not isinstance(arr, dndarray.DNDarray):
+            raise TypeError("arr must be a DNDarray, is currently a {}".format(type(self)))
+        if not isinstance(tiles_per_proc, int):
+            raise TypeError("tiles_per_proc must be an int, is currently a {}".format(type(self)))
+        if tiles_per_proc < 1:
+            raise ValueError("Tiles per process must be >= 1, currently: {}".format(tiles_per_proc))
+        if len(arr.shape) != 2:
+            raise ValueError("Arr must be 2 dimensional, current shape {}".format(arr.shape))
+
+        lshape_map = arr.create_lshape_map()
+
+        # if there is only one element of the diagonal on the next process
+        d = 1 if tiles_per_proc <= 2 else tiles_per_proc - 1
+        redist = torch.where(
+            torch.cumsum(lshape_map[..., arr.split], dim=0) >= arr.gshape[arr.split - 1] - d
+        )[0]
+        if redist.numel() > 0 and arr.gshape[0] > arr.gshape[1] and redist[0] != arr.comm.size - 1:
+            target_map = lshape_map.clone()
+            target_map[redist[0]] += d
+            target_map[redist[0] + 1] -= d
+            arr.redistribute_(lshape_map=lshape_map, target_map=target_map)
+
+        row_per_proc_list = [tiles_per_proc] * arr.comm.size
+
+        last_diag_pr, col_per_proc_list, col_inds, tile_columns = _create_cols_square_diag(
+            arr, lshape_map, tiles_per_proc
+        )
+
+        # need to adjust the lshape if the splits overlap...
+        if arr.split == 0 and tiles_per_proc == 1:
```

Comment:
The if conditions here seem highly nested. Can we break them up into sub functions/a function map?